include/mrdox/Support/SharedPtr.hpp

```diff
@@ -0,0 +1,325 @@
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// Copyright (c) 2023 Vinnie Falco (vinnie.falco@gmail.com)
+//
+// Official repository: https://github.com/cppalliance/mrdox
+//
+
+#ifndef MRDOX_API_SUPPORT_SHAREDPTR_HPP
+#define MRDOX_API_SUPPORT_SHAREDPTR_HPP
+
+#include <mrdox/Platform.hpp>
+#include <atomic>
+
+namespace clang {
+namespace mrdox {
+
+template<class T>
+class SharedPtr;
+
+template<class T>
+class AtomicSharedPtr;
+
+template<class U, class... Args>
+SharedPtr<U>
+makeShared(Args&&...);
+
+//------------------------------------------------
+
+class SharedPtrBase
+{
+protected:
+    struct Base
+    {
+        virtual ~Base() = default;
+        Base() noexcept = default;
+        Base* addref() noexcept;
+        void release() noexcept;
+        virtual void* get() noexcept = 0;
+        std::atomic<std::size_t> refs = 0;
+    };
+
+    template<class U, class... Args>
+    friend SharedPtr<U>
+    makeShared(Args&&...);
+};
+
+inline
+auto
+SharedPtrBase::
+Base::
+addref() noexcept ->
+    Base*
+{
+    ++refs;
+    return this;
+}
+
+inline
+void
+SharedPtrBase::
+Base::
+release() noexcept
+{
+    if(--refs > 0)
+        return;
+    delete this;
+}
+
+//------------------------------------------------
+
+/** A simple smart pointer container with shared ownership.
+*/
+template<class T>
+class SharedPtr : private SharedPtrBase
+{
+    Base* p_ = nullptr;
+
+    template<class U>
+    friend class SharedPtr;
+
+    template<class U>
+    friend class AtomicSharedPtr;
+
+    SharedPtr(Base*) noexcept;
+
+public:
+    ~SharedPtr();
+    SharedPtr() noexcept = default;
+    SharedPtr(SharedPtr const& other) noexcept;
+    SharedPtr& operator=(SharedPtr const& other) noexcept;
```

Comment:
Do you need this copy constructor and assignment operator? Could they removed in favour of the templated ones below?

Reply:
Its a good question. I am not sure. Peter might know :)